Added a best practices for feature routing

akilisosa opened the pull request:
docs: add best practices recommendation for feature routing to style guide



## PR Checklist
Please check if your PR fulfills the following requirements:

- [ x ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ x ] Tests for the changes have been added (for bug fixes / features)
- [ x ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [x] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ x ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Currently the angular style guide does not include documentation on best practices for routing. 
There are suggestions that can be found here https://angular.io/guide/router but they are not reflected in the styleGuide. 

Issue Number: N/A


## What is the new behavior?
It includes some basic suggestions for app routing. It states that the app routing module should mirror the app module to reduce clutter in that module, and that it also states that you shouldn't use routerLink='..' because that could lead to an unstable path. 

## Does this PR introduce a breaking change?
This might have a typo. 
- [ ] Yes
- [ x ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information 
Its late, if there are typos i am sorry.

akilisosa committed:
Added a best practices for feature routing

I suggested that feature routing module mirror  the regular feature module, this helps reduce clutter of code in the feature module.
I also suggested that people don't use routerLink='..' because that might lead to an unstable route. 

I also fixed one typo that i found by accident, one consider was marked with an avoid class.

kapunahelewong commented:
Thank you for this, @akilisosa! While we still need some further reviews from engineering and docs, you'll ultimately want to amend your commit message so that it passes.

Try this:
```
docs: add best practices recommendation for feature routing to style guide
```

Here are my own notes on [amending commit messages](https://gist.github.com/kapunahelewong/567c30f7b353fd1c39bbf6d8343e2dca#how-to-amend-a-commit). Let me know when you're ready or if you have any questions during the process.

akilisosa commented:
Thank you so much @kapunahelewong . This is my first time making a contribution like this so its very exciting! 

When reviewing the documentation, I was navigated to a page called AngularJS git commit message conventions. I was wondering if this is correct or if its suppose to be linked to another document.

jbogarthyde commented on aio/content/guide/styleguide.md:
```diff
@@ -2452,11 +2452,97 @@ area, folder, and file; for example, `app/heroes/heroes.module.ts` defines `Hero
+
+
+
+**Do** create an route for all distinct features in an application;
```
...create **a** route for all distinct features in an application. For example...

(make this two sentences)

jbogarthyde commented on aio/content/guide/styleguide.md:
```diff
@@ -2452,11 +2452,97 @@ area, folder, and file; for example, `app/heroes/heroes.module.ts` defines `Hero
+
+
+**Do** create an route for all distinct features in an application;
+for example, the `crisis-center-routing.module.ts`  helps navitagate through the different views in that feature.
```
navigate (spelling)

akilisosa replied:
Hey @jbogarthyde I am very new to this process. Am I suppose to be making these changes or are you reviewing these and making the changes?

jbogarthyde replied:
Please make the requested changes. If I make changes in your PR, git gets confused about different contributors.

jbogarthyde commented on aio/content/guide/styleguide.md:
```diff
@@ -2452,11 +2452,97 @@ area, folder, and file; for example, `app/heroes/heroes.module.ts` defines `Hero
+
+
+**Consider** The complexity of your application. The routing module should mirror naming conventions and location of feature modules. If you are working on a small application do not worry about this.
+
```
**Consider** The complexity of your application. In a larger application, the routing module should mirror your naming conventions and the locations of feature modules.

jbogarthyde commented on aio/content/guide/styleguide.md:
```diff
@@ -2452,11 +2452,97 @@ area, folder, and file; for example, `app/heroes/heroes.module.ts` defines `Hero
+
+
+
+**Avoid** defining long paths ie. path: 'home/crisis-center/crisis-center-list/crisis-list-detail:3'.
```
... long paths, such as `path: 'home/crisis-center/crisis-center-list/crisis-list-detail:3'`. 
(use code font for code, no latin abbreviations)

jbogarthyde commented on aio/content/guide/styleguide.md:
```diff
@@ -2452,11 +2452,97 @@ area, folder, and file; for example, `app/heroes/heroes.module.ts` defines `Hero
+
+
+
+**Why?** Use of the routing module helps reduce clutter in the root and feature module files. 
```
Ii don't get the context here. Should you avoid long paths IN the routing module? Or do you avoid long paths BY USING the routing module?

akilisosa replied:
I think the confusion may have arose because i was not following Do, avoid, why structure. I am still confused about how to use this properly. 

In my experience it seems like you should just avoid writing long paths in general. It also was commented on a video i was watching that the order in which you define your routes plays a role. I'm not to sure about this, but i know that when i created a bunch of long paths, and used `routerLink='..'` I ran into problems. If this is no longer a problem in Angular 9 please let me know.

jbogarthyde commented on aio/content/guide/styleguide.md:
```diff
@@ -2452,11 +2452,97 @@ area, folder, and file; for example, `app/heroes/heroes.module.ts` defines `Hero
+
+
+
+**Do** use routerLink with the specified component name to navigate to different components.
```
`routerLink` (code font)
 
also, how about a positive example?

jbogarthyde replied:
(still not using code font)
Do you mean "Set `routerLink` to a specific component, rather than a relative path." ?

jbogarthyde commented on aio/content/guide/styleguide.md:
```diff
@@ -2452,11 +2452,97 @@ area, folder, and file; for example, `app/heroes/heroes.module.ts` defines `Hero
+
+
+**Avoid** routerLink='..'.
+
```
Didn't you just say you SHOULD use it?

akilisosa replied:
I feel like this may have crossed into the ordering thing you were telling me about so that it can all make sense. Do, Avoid, Why. or Do, Why, Avoid.

jbogarthyde commented on aio/content/guide/styleguide.md:
```diff
@@ -2452,11 +2452,97 @@ area, folder, and file; for example, `app/heroes/heroes.module.ts` defines `Hero
+
+
+
+**Why?**  Specifying the routerLink location helps stabilize navigation.   
```
Is this a reason to use it or avoid it?
(also, `routerLink`)

jbogarthyde reviewed (changes requested):
Marked some typos-- and also needs some clarification.

akilisosa committed:
Update styleguide.md

I made the changes that were suggested to me by Judy Bogart. I also tried adding a new section so that it makes more sense, but it probably needs to be reviewed again. also maybe the particular bug that has to do long routes and their order has been fixed in angular 9.

akilisosa committed:
Update StyleGuide

fixed typos

mary-poppins commented:
You can preview c8a8f87 at https://pr34805-c8a8f87.ngbuilds.io/.

jbogarthyde commented on aio/content/guide/styleguide.md:
```diff
@@ -2452,11 +2452,116 @@ area, folder, and file; for example, `app/heroes/heroes.module.ts` defines `Hero
+
+
+**Why?** Use of the routing module helps increase readability and stabilization of your code. The order in which you define your paths effect how the application runs. When you add many complex paths in one location you may run into bugs.
+
```
effect -> affect

jbogarthyde commented on aio/content/guide/styleguide.md:
```diff
@@ -2452,11 +2452,116 @@ area, folder, and file; for example, `app/heroes/heroes.module.ts` defines `Hero
+
+**Do** use routerLink with the specified component name to navigate to different components.
+`<a routerLink="/crisis-center" routerLinkActive="active">Crisis Center</a>`
+
```
For example:

jbogarthyde commented on aio/content/guide/styleguide.md:
```diff
@@ -2452,11 +2452,116 @@ area, folder, and file; for example, `app/heroes/heroes.module.ts` defines `Hero
+
+
+**Avoid** The use of the router links to relative paths. `routerLink='..'` 
+
```
Why? Is this really a best practice, or have you just had trouble with it?

akilisosa replied:
I don't know. I did not see a relative path used in any of the angular examples, so I think according to the team that developed the examples it is a best practice. The convince of a relative path is great. Maybe I should change **Do** use routerLink with the specified components... to a **Consider**, and remove this line.

jelbourn replied:
I don't believe the examples were intended to be normative in this way; I don't know of any strong reason to discourage relative paths.

jbogarthyde commented on aio/content/guide/styleguide.md:
```diff
@@ -2452,11 +2452,116 @@ area, folder, and file; for example, `app/heroes/heroes.module.ts` defines `Hero
+
+
+
+**Do** Use app routing to define short paths that relate to the components. 
```
Use -> use

jbogarthyde commented on aio/content/guide/styleguide.md:
```diff
@@ -2452,11 +2452,116 @@ area, folder, and file; for example, `app/heroes/heroes.module.ts` defines `Hero
+
+
+
+**Avoid** The use of the router links to relative paths. `routerLink='..'` 
```
perhaps: 

**Avoid**  using router links to non-specific relative paths (`routerLink='..'` ).

akilisosa replied:
I added a **Why** instead of changing the **Do** to a **Consider** .

akilisosa committed:
Updated StyleGuide.md

made requested changes, and added a why to the bottom.

akilisosa committed:
Update styleGuide.md

added a why, and made other requested changes.

akilisosa committed:
updated styleguide.md

removed bias

jelbourn commented on aio/content/guide/styleguide.md:
```diff
@@ -2452,11 +2452,125 @@ area, folder, and file; for example, `app/heroes/heroes.module.ts` defines `Hero
+
+
+**Do** create a route for all distinct features in an application.
+For example, the `crisis-center-routing.module.ts`  helps navigate through the different views in that feature.
```
I'm not sure Angular should have an opinion on the information architecture of applications. Ultimately the "best" organization of routes is going to vary from app to app, so it may be better to leave this area unopinionated.

mgechev replied:
I generally agree that it's application specific. At the same time think it's a good idea to encourage code splitting. Not sure if that's the right way to go.

akilisosa replied:
thank you guys for reviewing this. I am still a little new to developing and I am not sure what informational architecture refers to specifically, if it refers to the actual path names, I agree that shouldn't be angular's opinion. 

If it is referring to the folder structure: the angular style guide comments on that when it is talking about FLAT principles - keeping folders with up to 7 items, and how you should just group components together, and change names of things so they are easy to locate. 

I think when I originally proposed this I was considering this as an extension of those thoughts. 

Again I am excited to learn and be participating with you all in this way. Thank you so much.

jelbourn commented on aio/content/guide/styleguide.md:
```diff
@@ -2452,11 +2452,125 @@ area, folder, and file; for example, `app/heroes/heroes.module.ts` defines `Hero
+
+
+
+**Consider** the complexity of your application. In a larger application, the routing module should mirror your naming conventions and the locations of feature modules.
```
We actually want to move _away_ from having/recommending a routing module based on feedback that it was unnecessarily complicated.

jelbourn commented on aio/content/guide/styleguide.md:
```diff
@@ -2452,11 +2452,125 @@ area, folder, and file; for example, `app/heroes/heroes.module.ts` defines `Hero
+
+
+**Do** use app routing to define short paths that relate to the components. 
+For example: `path: 'crisis-center' `
```
Similar to the earlier comment, I'm not sure that Angular should have an opinion about route names themselves.

jelbourn commented on aio/content/guide/styleguide.md:
```diff
@@ -2452,11 +2452,125 @@ area, folder, and file; for example, `app/heroes/heroes.module.ts` defines `Hero
+
+
+
+**Do** use routerLink with the specified component name to navigate to different components.
```
I don't know that use of `routerLink` is something that belongs in the style guide, since it's the primary mechanism through with `@angular/router` functions, rather than a style choice / best practice.

jelbourn reviewed (commented):
Hi @akilisosa, thanks for taking the time to send a PR

I'm not sure the changes proposed here align with recommendations we would want to make. + @mgechev to fact-check me on this

mgechev commented:
I agree with Jeremy. I like some of the ideas, especially around code-splitting. Recently we had a discussion to move away from the routing module recommendation to simplify application structure.

I'd suggest to close the PR, but continue the discussion in a separate issue. See https://github.com/angular/angular/issues/37351

akilisosa commented:
This point may have gotten lost in the mix:

The main reason I suggested this pull request was because I was part of a team that developed a large application and used one centralized routing system and relative paths. The behavior became increasingly unstable until I re-organized the project with the use of the routing table.  I don't know if there is still an issue with the keeping all of your paths in one centralized place but it was an issue at the time.

mgechev commented:
@akilisosa IMO, it's worth it to mention the decentralized routing suggestion. If we can shrink the scope, it'll be easier to merge.

akilisosa commented:
@mgechev sorry. I think i missed the point of this document, and what I was contributing to. If I was having issues I could have raised an issue with that aspect right? 

Just let me know what you all want me to do, and how to contribute.

jelbourn commented:
I think it would be easiest to close this PR, and then open a new issue for discussing any more focused changes you might propose for the style guide. As @mgechev mentioned, there's potentially merit in adding guidance on decentralizing route configuration in some cases.

akilisosa commented:
Great! so do click on close and comment and thats it?

jelbourn commented:
Yep
